File: spring-docs-viewer/src/main/java/com/docs/viewer/common/file/service/FileService.java
package com.docs.viewer.common.file.service;

import com.docs.viewer.common.file.dto.FileAttachInfo;
import com.docs.viewer.common.file.dto.FileInfo;
import com.docs.viewer.common.file.entity.File;
import com.docs.viewer.common.file.repository.FileRepository;
import com.docs.viewer.global.common.setting.FileSetting;
import com.docs.viewer.global.common.utils.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import jakarta.transaction.Transactional;
import java.io.IOException;
import java.nio.file.Path;
import java.util.List;
import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
public class FileService {

    private static final String FILE_DIRECTORY = "cmmn";

    private final FileRepository fileRepository;
    private final FileSetting fileSetting;

    @Transactional
    public byte[] findFile(Integer attachFile) throws Exception {
        Optional<File> info = this.fileRepository.findById(attachFile.longValue());
        if (info.isPresent()) {
            File file = info.get();
            if('N' == file.getCloudYn()) {
                String rootPath = fileSetting.getFilePath();
                String filePath = file.getFilePath().replace("..", rootPath);
                return CryptoUtil.decryptFile(filePath);
            }
        }
        throw new IOException("해당되는 파일을 찾을 수 없습니다.");
    }

    @Transactional
    public FileInfo saveFile(MultipartFile file, String targetFolder) throws Exception {
        String rootPath = fileSetting.getFilePath();
        Path uploadDirectory = FileUtil.getUploadDirectory(FileUtil.getDirectory(rootPath, targetFolder));
        String fileName = file.getOriginalFilename();
        String contentType = file.getContentType();
        Integer fileSize = Integer.valueOf(String.valueOf(file.getSize()));
        File fileInfo = this.fileRepository.save(File.of(fileName, FileUtil.getFilePath(uploadDirectory), contentType, fileSize, 'N'));
        CryptoUtil.encryptFile(file.getBytes(), FileUtil.getTargetFile(fileInfo.getFilePath()));  // 파일 저장
        return FileInfo.of(fileInfo);
    }

    @Transactional
    public List<FileAttachInfo> saveFile(List<MultipartFile> files) throws Exception {
        List<FileInfo> fileInfoList = files.stream().map(ExceptionUtil.wrapFunction(file -> this.saveFile(file, FILE_DIRECTORY))).toList();
        return fileInfoList.stream().map(FileAttachInfo::of).toList();
    }
}

File: dotnet-docs-viewer/Services/FileService.cs
using Common.Utils;
using DB.Model;
using Microsoft.Extensions.Options;
using WebApp.Dto;
using WebApp.Repository;

namespace WebApp.Service
{
    public interface IFileService
    {
        Task<byte[]> FindFileAsync(int attachFile);
        Task<List<FileAttachInfoDto>> SaveFileAsync(List<IFormFile> files);
    }
    
    public class FileService : IFileService
    {
        private static readonly string FileDirectory = "cmmn";
            
        private readonly IRepository<FileInfoBaseModel> _fileRepository;
        private readonly IOptions<FileSetting> _fileSetting;

        public FileService(IRepository<FileInfoBaseModel> fileRepository, IOptions<FileSetting> fileSetting)
        {
            _fileRepository = fileRepository;
            _fileSetting = fileSetting;
        }
        
        public async Task<byte[]> FindFileAsync(int attachFile)
        {
            var file = await _fileRepository.GetByIdAsync(attachFile);

            if (file == null)
                throw new FileNotFoundException("해당되는 파일을 찾을 수 없습니다.");

            if (file.CloudYn == 'N')
            {
                var rootPath = _fileSetting.Value.FilePath;
                var filePath = file.FilePath.Replace("..", rootPath);

                return CryptoUtil.DecryptFile(filePath);
            }

            throw new InvalidOperationException("Cloud 파일은 아직 지원하지 않습니다.");
        }
        
        public async Task<List<FileAttachInfoDto>> SaveFileAsync(List<IFormFile> files)
        {
            var result = new List<FileAttachInfoDto>();

            foreach (var file in files)
            {
                var info = await SaveFileAsync(file, FileDirectory);
                result.Add(FileAttachInfoDto.Of(info));
            }

            return result;
        }
        
        private async Task<FileInfoDto> SaveFileAsync(IFormFile file, string targetFolder)
        {
            var rootPath = _fileSetting.Value.FilePath;

            var uploadDirectory = FileUtil.GetUploadDirectory(
                FileUtil.GetDirectory(rootPath, targetFolder)
            );

            var filePath = FileUtil.GetFilePath(uploadDirectory);

            var fileInfo = FileInfoModel.Of(
                file.FileName,
                filePath,
                file.ContentType,
                (int)file.Length,
                'N'
            );

            await _fileRepository.SaveAsync(fileInfo);

            var fileBytes = await ToByteArrayAsync(file);
            CryptoUtil.EncryptFile(fileBytes, FileUtil.GetTargetFile(fileInfo.FilePath));

            return FileInfoDto.Of(fileInfo);
        }
        
        private static async Task<byte[]> ToByteArrayAsync(IFormFile file)
        {
            await using var ms = new MemoryStream();
            await file.CopyToAsync(ms);
            return ms.ToArray();
        }
    }
}